src/proof/theory_proof.cpp

```diff
@@ -1264,4 +1263,387 @@ void TheoryProof::printRewriteProof(std::ostream& os, const Node &n1, const Node
   os << "))";
 }
 
+
+inline bool TheoryProof::match(TNode n1, TNode n2, theory::TheoryId theoryId)
+{
+  bool ufProof = (theoryId == theory::THEORY_UF);
+  Debug(ufProof ? "pf::uf" : "mgd") << "match " << n1 << " " << n2 << std::endl;
+  if (ProofManager::currentPM()->hasOp(n1))
+  {
+    n1 = ProofManager::currentPM()->lookupOp(n1);
+  }
+  if (ProofManager::currentPM()->hasOp(n2))
+  {
+    n2 = ProofManager::currentPM()->lookupOp(n2);
+  }
+  Debug(ufProof ? "pf::uf" : "mgd") << "+ match " << n1 << " " << n2
+                                    << std::endl;
+  if (!ufProof)
+  {
+    Debug("pf::array") << "+ match: step 1" << std::endl;
+  }
+  if (n1 == n2)
+  {
+    return true;
+  }
+
+  if (n1.getType().isFunction() && n2.hasOperator())
+  {
+    if (ProofManager::currentPM()->hasOp(n2.getOperator()))
+    {
+      return n1 == ProofManager::currentPM()->lookupOp(n2.getOperator());
+    }
+    else
+    {
+      return n1 == n2.getOperator();
+    }
+  }
+
+  if (n2.getType().isFunction() && n1.hasOperator())
+  {
+    if (ProofManager::currentPM()->hasOp(n1.getOperator()))
+    {
+      return n2 == ProofManager::currentPM()->lookupOp(n1.getOperator());
+    }
+    else
+    {
+      return n2 == n1.getOperator();
+    }
+  }
+
+  if (n1.hasOperator() && n2.hasOperator()
+      && n1.getOperator() != n2.getOperator())
+  {
+    if (ufProof
+        || !((n1.getKind() == kind::SELECT
+              && n2.getKind() == kind::PARTIAL_SELECT_0)
+             || (n1.getKind() == kind::SELECT
+                 && n2.getKind() == kind::PARTIAL_SELECT_1)
+             || (n1.getKind() == kind::PARTIAL_SELECT_1
+                 && n2.getKind() == kind::SELECT)
+             || (n1.getKind() == kind::PARTIAL_SELECT_1
+                 && n2.getKind() == kind::PARTIAL_SELECT_0)
+             || (n1.getKind() == kind::PARTIAL_SELECT_0
+                 && n2.getKind() == kind::SELECT)
+             || (n1.getKind() == kind::PARTIAL_SELECT_0
+                 && n2.getKind() == kind::PARTIAL_SELECT_1)))
+    {
+      return false;
+    }
+  }
+
+  for (size_t i = 0; i < n1.getNumChildren() && i < n2.getNumChildren(); ++i)
+  {
+    if (n1[i] != n2[i])
+    {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+void TheoryProof::assertAndPrint(
+    const theory::eq::EqProof& pf,
+    const ProofLetMap& map,
+    const theory::TheoryId theoryId,
+    int* neg,
```

Comment:
Maybe return the new value of `neg` instead of using a pointer? It also looks like `neg` is always -1 when `assertAndPrint` is called. Maybe this doesn't even need to be an argument?